### PR TITLE
Add EnumeratedArray type

### PR DESCRIPTION
**Pull request by litherum**
#### 66dcc4e97851c90a6e7d7162a27686a9802ea1a7
<pre>
Add EnumeratedArray type
<a href="https://bugs.webkit.org/show_bug.cgi?id=241451">https://bugs.webkit.org/show_bug.cgi?id=241451</a>

Reviewed by Cameron McCormack.

This is an std::array where the indices of the array are values of an enum (rather than a size_t).
This assumes the values of the enum start at 0 and monotonically increase by 1
(so the conversion function between size_t and the enum is just a simple static_cast).

EnumeratedArray includes almost all of the functions from std::array, except for a few which I
intentionally omitted.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/EnumeratedArray.h: Added.
(WTF::EnumeratedArray::EnumeratedArray):
(WTF::EnumeratedArray::operator=):
(WTF::EnumeratedArray::at):
(WTF::EnumeratedArray::at const):
(WTF::EnumeratedArray::operator[]):
(WTF::EnumeratedArray::operator[] const):
(WTF::EnumeratedArray::front):
(WTF::EnumeratedArray::front const):
(WTF::EnumeratedArray::back):
(WTF::EnumeratedArray::back const):
(WTF::EnumeratedArray::fill):
(WTF::EnumeratedArray::operator== const):
(WTF::EnumeratedArray::operator!= const):
(WTF::EnumeratedArray::operator&lt; const):
(WTF::EnumeratedArray::operator&lt;= const):
(WTF::EnumeratedArray::operator &gt; const):
(WTF::EnumeratedArray::operator &gt;= const):
(WTF::EnumeratedArray::index):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/EnumeratedArray.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/251453@main">https://commits.webkit.org/251453@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295447">https://svn.webkit.org/repository/webkit/trunk@295447</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/WTF/wtf/EnumeratedArray.h [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    template <typename Key2, typename T2, Key2 N2>
+    bool operator!=(const EnumeratedArray<Key2, T2, N2>& rhs) const
+    {
+        return m_storage != rhs.m_storage;
+    }
+
+    template <typename Key2, typename T2, Key2 N2>
+    bool operator<(const EnumeratedArray<Key2, T2, N2>& rhs) const
```
Comment (heycam) [inline Source/WTF/wtf/EnumeratedArray.h]: TIL `std::array` has an `operator<`.

File: Source/WTF/wtf/EnumeratedArray.h [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+#include <array>
+
+namespace WTF {
+
+// This is an std::array where the indices of the array are values of an enum (rather than a size_t).
+// This assumes the values of the enum start at 0 and monotonically increase by 1
+// (so the conversion function between size_t and the enum is just a simple static_cast).
+// N is the maximum value of the enum, which determines the size of the array.
+template <typename Key, typename T, Key N>
```
Comment (heycam) [inline Source/WTF/wtf/EnumeratedArray.h]: Might be slightly confusing that `N` here is not the count, but the last value. Maybe name this `LastValue`?

It's also different from `std::array` which does take a count, but that can't be helped if we don't want to add a final `Count` enum value to our enums, and we assume that writing things like `EnumeratedArray<MyEnum, ..., static_cast<MyEnum>(static_cast<size_t>(MyEnum::LastValue) + 1)>` is too awkward (which it is).

**Review by heycam — APPROVED** [review]
I think it'd be fine to have a `size()` member that just returns `size_t` rather than `size_type`.

File: Source/WTF/wtf/EnumeratedArray.h (hunk 2 above)
Comment (litherum) [reply Source/WTF/wtf/EnumeratedArray.h]: Yeah, you're exactly right.

**Pull request merged**